Annotation can be applied either on class or its methods

### Describe the bug

When upgrading from quarkus 3.29.4 to 3.30.0 the error below occurs when building the application. I strongly suspect the following change in 3.30.0: https://github.com/quarkusio/quarkus/issues/49474


application.properties
```
quarkus.openapi-generator.codegen.input-base-dir=target/openapi
quarkus.rest-client.client-prefix.connect-timeout=10000
quarkus.openapi-generator.codegen.enable-security-generation=false

auth.tokenPropagation=@io.quarkus.oidc.token.propagation.common.AccessToken
generator.modelAnnotations=@lombok.experimental.SuperBuilder(toBuilder = true);@lombok.EqualsAndHashCode;@lombok.AllArgsConstructor;@lombok.NoArgsConstructor

quarkus.openapi-generator.codegen.spec.openapi_file_yaml.additional-model-type-annotations=${generator.modelAnnotations}
quarkus.openapi-generator.codegen.spec.openapi_file_yaml.base-package=nl.organisation.pack.application
quarkus.openapi-generator.codegen.spec.openapi_file_yaml.additional-api-type-annotations=${auth.tokenPropagation}
quarkus.openapi-generator.codegen.spec.openapi_file_yaml.mutiny=true
quarkus.openapi-generator.openapi_file_yaml.auth.bearerAuth.token-propagation=true
```


The interface that gets generated by the openapi generator config above looks like this:

```
@Path("/foo")
@RegisterRestClient(
    baseUri = "http://localhost/api",
    configKey = "openapi_file_yaml"
)
@GeneratedClass(
    value = "openapi_file.yaml",
    tag = "bar"
)
@ApplicationScoped
@AccessToken
public interface ServiceApi {
    @OperationMarkers({@OperationMarker(
    name = "bearerAuth",
    openApiSpecId = "openapi_file_yaml",
    operationId = "foo",
    method = "GET",
    path = "/foo"
), @OperationMarker(
    name = "OAuth",
    openApiSpecId = "openapi_file_yaml",
    operationId = "foo",
    method = "GET",
    path = "/foo"
)})
    @GET
    @Produces({"application/json", "application/problem+json"})
    @GeneratedMethod("foo")
    Uni<Foo> fooBar();
}

```


### Expected behavior

The class should get generated, as the `@AccessToken` annotation is only on the interface. Right now I throws an error

### Actual behavior

This error gets thrown:
```
69255 [INFO] 
69255 [INFO] --- surefire:3.5.4:test (default-test) @ application-name ---
69310 [INFO] Surefire report directory: /builds/organisation/platforms/launchpad/applications/application/app/application-name/target/surefire-reports
69312 [INFO] Using auto detected provider org.apache.maven.surefire.junitplatform.JUnitPlatformProvider
69326 [INFO] 
69326 [INFO] -------------------------------------------------------
69326 [INFO]  T E S T S
69326 [INFO] -------------------------------------------------------
SLF4J(W): Class path contains multiple SLF4J providers.
SLF4J(W): Found provider [org.slf4j.impl.JBossSlf4jServiceProvider@1bb40695]
SLF4J(W): Found provider [org.slf4j.jul.JULServiceProvider@7cde748f]
SLF4J(W): See https://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J(I): Actual provider is of type [org.slf4j.impl.JBossSlf4jServiceProvider@1bb40695]
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.oidc.token.propagation.common.deployment.OidcTokenPropagationCommonProcessor#collectAccessTokenInstances threw an exception: java.lang.RuntimeException: io.quarkus.oidc.token.propagation.common.AccessToken annotation can be applied either on class nl.organisation.pack.application.api.ServiceApi or its methods
	at io.quarkus.oidc.token.propagation.common.deployment.OidcTokenPropagationCommonProcessor.lambda$collectAccessTokenInstances$1(OidcTokenPropagationCommonProcessor.java:58)
	at java.base/java.util.HashMap.forEach(HashMap.java:1429)
	at io.quarkus.oidc.token.propagation.common.deployment.OidcTokenPropagationCommonProcessor.collectAccessTokenInstances(OidcTokenPropagationCommonProcessor.java:56)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:872)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1694)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:388)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:305)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:61)
	at io.quarkus.test.junit.AppMakerHelper.getStartupAction(AppMakerHelper.java:253)
	at io.quarkus.test.junit.classloading.FacadeClassLoader.getOrCreateRuntimeClassLoader(FacadeClassLoader.java:589)
	at io.quarkus.test.junit.classloading.FacadeClassLoader.getQuarkusClassLoader(FacadeClassLoader.java:486)
	at io.quarkus.test.junit.classloading.FacadeClassLoader.loadClass(FacadeClassLoader.java:368)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:536)
	at java.base/java.lang.Class.forName(Class.java:515)
	at org.junit.platform.commons.util.ReflectionUtils.lambda$tryToLoadClass$10(ReflectionUtils.java:886)
	at org.junit.platform.commons.function.Try.lambda$call$0(Try.java:57)
	at org.junit.platform.commons.function.Try.of(Try.java:93)
	at org.junit.platform.commons.function.Try.call(Try.java:57)
	at org.junit.platform.commons.util.ReflectionUtils.tryToLoadClass(ReflectionUtils.java:873)
	at org.junit.platform.commons.util.ReflectionUtils.tryToLoadClass(ReflectionUtils.java:810)
	at org.junit.platform.commons.support.ReflectionSupport.tryToLoadClass(ReflectionSupport.java:94)
	at org.junit.platform.engine.discovery.ClassSelector.getJavaClass(ClassSelector.java:95)
	at org.junit.jupiter.engine.discovery.ClassSelectorResolver.resolve(ClassSelectorResolver.java:87)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.lambda$resolve$2(EngineDiscoveryRequestResolution.java:135)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1685)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolve(EngineDiscoveryRequestResolution.java:189)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolve(EngineDiscoveryRequestResolution.java:126)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolveCompletely(EngineDiscoveryRequestResolution.java:92)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.run(EngineDiscoveryRequestResolution.java:83)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolver.resolve(EngineDiscoveryRequestResolver.java:148)
	at org.junit.jupiter.engine.discovery.DiscoverySelectorResolver.resolveSelectors(DiscoverySelectorResolver.java:63)
	at org.junit.jupiter.engine.JupiterTestEngine.discover(JupiterTestEngine.java:70)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:195)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverSafely(EngineDiscoveryOrchestrator.java:174)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:119)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:84)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:104)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:83)
	at org.junit.platform.launcher.core.DelegatingLauncher.discover(DelegatingLauncher.java:42)
	at org.junit.platform.launcher.core.InterceptingLauncher.lambda$discover$0(InterceptingLauncher.java:33)
	at org.junit.platform.launcher.core.ClasspathAlignmentCheckingLauncherInterceptor.intercept(ClasspathAlignmentCheckingLauncherInterceptor.java:25)
	at org.junit.platform.launcher.core.InterceptingLauncher.discover(InterceptingLauncher.java:33)
	at org.junit.platform.launcher.core.DelegatingLauncher.discover(DelegatingLauncher.java:42)
	at org.apache.maven.surefire.junitplatform.LauncherAdapter.discover(LauncherAdapter.java:47)
	at org.apache.maven.surefire.junitplatform.TestPlanScannerFilter.accept(TestPlanScannerFilter.java:51)
	at org.apache.maven.surefire.api.util.DefaultScanResult.applyFilter(DefaultScanResult.java:87)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.scanClasspath(JUnitPlatformProvider.java:156)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:136)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.oidc.token.propagation.common.deployment.OidcTokenPropagationCommonProcessor#collectAccessTokenInstances threw an exception: java.lang.RuntimeException: io.quarkus.oidc.token.propagation.common.AccessToken annotation can be applied either on class nl.organisation.pack.application.api.ServiceApi or its methods
at io.quarkus.oidc.token.propagation.common.deployment.OidcTokenPropagationCommonProcessor.lambda$collectAccessTokenInstances$1(OidcTokenPropagationCommonProcessor.java:58)
	at java.base/java.util.HashMap.forEach(HashMap.java:1429)
	at io.quarkus.oidc.token.propagation.common.deployment.OidcTokenPropagationCommonProcessor.collectAccessTokenInstances(OidcTokenPropagationCommonProcessor.java:56)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:872)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1694)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
	at io.quarkus.builder.Execution.run(Execution.java:139)
	at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:78)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:160)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:384)
	... 55 more
```

### How to Reproduce?

An quarkus application with version >= 3.30.0

And an open-api generator with these annotations:
```
quarkus.openapi-generator.codegen.enable-security-generation=false
quarkus.openapi-generator.codegen.spec.openapi_file_yaml.additional-api-type-annotations=@io.quarkus.oidc.token.propagation.common.AccessToken
```


### Output of `uname -a` or `ver`

Darwin OMD231914 24.6.0 Darwin Kernel Version 24.6.0: Wed Nov  5 21:33:58 PST 2025; root:xnu-11417.140.69.705.2~1/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk 21.0.2 2024-01-16 LTS OpenJDK Runtime Environment Corretto-21.0.2.13.1 (build 21.0.2+13-LTS) OpenJDK 64-Bit Server VM Corretto-21.0.2.13.1 (build 21.0.2+13-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.30.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.12

### Additional information

The environment doesn't matter that much. It cannot build on my macbook with java21 or java25 and also not in my CI/CD pipeline with the mandrel builder image jdk-21 or jdk-25.

--- a/extensions/oidc-token-propagation-common/deployment/src/main/java/io/quarkus/oidc/token/propagation/common/deployment/AccessTokenRequestFilterGenerator.java
+++ b/extensions/oidc-token-propagation-common/deployment/src/main/java/io/quarkus/oidc/token/propagation/common/deployment/AccessTokenRequestFilterGenerator.java
@@ -2,8 +2,10 @@ package io.quarkus.oidc.token.propagation.common.deployment;
 
 import java.lang.annotation.RetentionPolicy;
 import java.lang.reflect.Modifier;
+import java.util.Arrays;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.random.RandomGenerator;
 
 import jakarta.annotation.Priority;
 import jakarta.inject.Singleton;
@@ -24,7 +26,7 @@ public final class AccessTokenRequestFilterGenerator {
 
     private static final int AUTHENTICATION = 1000;
 
-    private record RequestFilterKey(String clientName, boolean exchangeTokenActivated, MethodInfo targetMethodInfo) {
+    private record RequestFilterKey(String clientName, boolean exchangeTokenActivated, MethodDescription methodDescription) {
     }
 
     private final BuildProducer<UnremovableBeanBuildItem> unremovableBeansProducer;
@@ -43,12 +45,12 @@ public final class AccessTokenRequestFilterGenerator {
     }
 
     public String generateClass(AccessTokenInstanceBuildItem instance) {
+        MethodDescription methodDescription = createMethodDescription(instance.getTargetMethodInfo());
         return cache.computeIfAbsent(
-                new RequestFilterKey(instance.getClientName(), instance.exchangeTokenActivated(),
-                        instance.getTargetMethodInfo()),
+                new RequestFilterKey(instance.getClientName(), instance.exchangeTokenActivated(), methodDescription),
                 i -> {
                     var adaptor = new GeneratedBeanGizmoAdaptor(generatedBeanProducer);
-                    String className = createUniqueClassName(i);
+                    String className = createUniqueClassName(i, instance);
                     try (ClassCreator classCreator = ClassCreator.builder()
                             .className(className)
                             .superClass(requestFilterClass)
@@ -77,21 +79,19 @@ public final class AccessTokenRequestFilterGenerator {
                          * return new MethodDescription(declaringClassName, methodName, parameterTypes);
                          * }
                          */
-                        if (i.targetMethodInfo != null) {
+                        if (i.methodDescription() != null) {
                             try (var methodCreator = classCreator.getMethodCreator("getMethodDescription",
                                     MethodDescription.class)) {
                                 methodCreator.addAnnotation(Override.class.getName(), RetentionPolicy.CLASS);
                                 methodCreator.setModifiers(Modifier.PROTECTED);
 
                                 // String methodName
-                                var methodName = methodCreator.load(i.targetMethodInfo.name());
+                                var methodName = methodCreator.load(i.methodDescription().getMethodName());
                                 // String declaringClassName
-                                var declaringClassName = methodCreator
-                                        .load(i.targetMethodInfo.declaringClass().name().toString());
+                                var declaringClassName = methodCreator.load(i.methodDescription().getClassName());
                                 // String[] paramTypes
                                 var paramTypes = methodCreator.marshalAsArray(String[].class,
-                                        i.targetMethodInfo.parameterTypes().stream()
-                                                .map(pt -> pt.name().toString()).map(methodCreator::load)
+                                        Arrays.stream(i.methodDescription().getParameterTypes()).map(methodCreator::load)
                                                 .toArray(ResultHandle[]::new));
                                 // new MethodDescription(declaringClassName, methodName, parameterTypes)
                                 var methodDescriptionCtor = MethodDescriptor.ofConstructor(MethodDescription.class,
@@ -112,11 +112,13 @@ public final class AccessTokenRequestFilterGenerator {
                 });
     }
 
-    private String createUniqueClassName(RequestFilterKey i) {
+    private String createUniqueClassName(RequestFilterKey i, AccessTokenInstanceBuildItem instance) {
         String uniqueClassName = "%s_%sClient_%sTokenExchange".formatted(requestFilterClass.getName(),
                 clientName(i.clientName()), exchangeTokenName(i.exchangeTokenActivated()));
-        if (i.targetMethodInfo != null) {
-            uniqueClassName = uniqueClassName + "_" + i.targetMethodInfo.name();
+        if (i.methodDescription != null) {
+            // we need the random so that we avoid conflicts between methods with the same name and different parameters
+            uniqueClassName += "_" + instance.getTargetMethodInfo().declaringClass().simpleName() + "_"
+                    + i.methodDescription.getMethodName() + "_" + RandomGenerator.getDefault().nextInt();
         }
         return uniqueClassName;
     }
@@ -135,5 +137,18 @@ public final class AccessTokenRequestFilterGenerator {
         } else {
             return "Default";
         }
+    }
+
+    private static MethodDescription createMethodDescription(MethodInfo mi) {
+        if (mi == null) {
+            return null;
+        }
+
+        String[] paramTypes = new String[mi.parametersCount()];
+        for (int i = 0; i < mi.parametersCount(); i++) {
+            paramTypes[i] = mi.parameterTypes().get(i).name().toString();
+        }
+        return new MethodDescription(mi.declaringClass().name().toString(), mi.name(),
+                paramTypes);
     }
 }

--- a/extensions/oidc-token-propagation-common/deployment/src/main/java/io/quarkus/oidc/token/propagation/common/deployment/OidcTokenPropagationCommonProcessor.java
+++ b/extensions/oidc-token-propagation-common/deployment/src/main/java/io/quarkus/oidc/token/propagation/common/deployment/OidcTokenPropagationCommonProcessor.java
@@ -2,8 +2,9 @@ package io.quarkus.oidc.token.propagation.common.deployment;
 
 import static java.util.stream.Collectors.groupingBy;
 
+import java.util.ArrayList;
 import java.util.List;
-import java.util.Objects;
+import java.util.stream.Collectors;
 
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationTarget;
@@ -20,7 +21,11 @@ public class OidcTokenPropagationCommonProcessor {
 
     @BuildStep
     public List<AccessTokenInstanceBuildItem> collectAccessTokenInstances(CombinedIndexBuildItem index) {
-        record ItemBuilder(AnnotationInstance instance) {
+        record ItemBuilder(AnnotationInstance instance, AnnotationTarget annotationTarget) {
+
+            ItemBuilder(AnnotationInstance instance) {
+                this(instance, instance.target());
+            }
 
             private String toClientName() {
                 var value = instance.value("exchangeTokenClient");
@@ -32,31 +37,40 @@ public class OidcTokenPropagationCommonProcessor {
             }
 
             private MethodInfo methodInfo() {
-                if (instance.target().kind() == AnnotationTarget.Kind.METHOD) {
-                    return instance.target().asMethod();
+                if (annotationTarget.kind() == AnnotationTarget.Kind.METHOD) {
+                    return annotationTarget.asMethod();
                 }
                 return null;
             }
 
             private String targetClassName() {
-                if (instance.target().kind() == AnnotationTarget.Kind.METHOD) {
-                    return instance.target().asMethod().declaringClass().name().toString();
+                if (annotationTarget.kind() == AnnotationTarget.Kind.METHOD) {
+                    return annotationTarget.asMethod().declaringClass().name().toString();
                 }
-                return instance.target().asClass().name().toString();
+                return annotationTarget.asClass().name().toString();
             }
 
             private AccessTokenInstanceBuildItem build() {
-                return new AccessTokenInstanceBuildItem(toClientName(), toExchangeToken(), instance.target(), methodInfo());
+                return new AccessTokenInstanceBuildItem(toClientName(), toExchangeToken(), annotationTarget, methodInfo());
             }
         }
         var accessTokenAnnotations = index.getIndex().getAnnotations(ACCESS_TOKEN);
-        var itemBuilders = accessTokenAnnotations.stream().map(ItemBuilder::new).toList();
+        var itemBuilders = accessTokenAnnotations.stream().map(ItemBuilder::new)
+                .collect(Collectors.toCollection(ArrayList::new));
         if (!itemBuilders.isEmpty()) {
             var targetClassToBuilders = itemBuilders.stream().collect(groupingBy(ItemBuilder::targetClassName));
             targetClassToBuilders.forEach((targetClassName, classBuilders) -> {
-                if (classBuilders.size() > 1 && classBuilders.stream().map(ItemBuilder::methodInfo).anyMatch(Objects::isNull)) {
-                    throw new RuntimeException(
-                            ACCESS_TOKEN + " annotation can be applied either on class " + targetClassName + " or its methods");
+                if (classBuilders.size() > 1) {
+                    var classLevelAnnotations = classBuilders.stream().filter(b -> b.methodInfo() == null).toList();
+                    if (!classLevelAnnotations.isEmpty()) {
+                        var classItemBuilder = classLevelAnnotations.get(0);
+                        // now we have @AccessToken on both class and method, so remove the class-level one
+                        // and apply it on all the methods not annotated itself
+                        itemBuilders.remove(classItemBuilder);
+                        var annotatedClass = classItemBuilder.instance.target().asClass();
+                        annotatedClass.methods().stream().filter(m -> !m.hasAnnotation(ACCESS_TOKEN))
+                                .map(mi -> new ItemBuilder(classItemBuilder.instance, mi)).forEach(itemBuilders::add);
+                    }
                 }
             });
         }

--- a/extensions/oidc-token-propagation-reactive/deployment/src/test/java/io/quarkus/oidc/token/propagation/reactive/deployment/test/AccessTokenAnnotationTest.java
+++ b/extensions/oidc-token-propagation-reactive/deployment/src/test/java/io/quarkus/oidc/token/propagation/reactive/deployment/test/AccessTokenAnnotationTest.java
@@ -41,7 +41,10 @@ public class AccessTokenAnnotationTest {
                             NamedClientDefaultExchange.class, MultiProviderFrontendResource.class, ProtectedResource.class,
                             CustomAccessTokenRequestFilter.class, NamedClientDefaultExchange_OnMethod.class,
                             DefaultClientEnabledExchange_OnMethod.class, DefaultClientDefaultExchange_OnMethod.class,
-                            MultipleClientsAndMultipleMethods.class)
+                            MultipleClientsAndMultipleMethods.class, DefaultClientDefaultExchange_RepeatedAnnotation.class,
+                            NamedClientDefaultExchange_RepeatedAnnotation.class,
+                            DefaultClientDefaultExchangeOnClass_NamedOnMethod.class,
+                            NamedClientDefaultExchangeOnClass_DefaultClientOnMethod.class)
                     .addAsResource(
                             new StringAsset(
                                     """
@@ -79,13 +82,21 @@ public class AccessTokenAnnotationTest {
         testRestClientTokenPropagation(true, "defaultClientEnabledExchange");
         testRestClientTokenPropagation(true, "defaultClientEnabledExchange_OnMethod");
         testRestClientTokenPropagation(true, "multipleClientsAndMultipleMethods_DefaultClientEnabledExchange");
+        testRestClientTokenPropagation(true,
+                "namedClientDefaultExchangeOnClassDefaultClientOnMethod_methodLevelDefaultClientEnabledExchange");
+        testRestClientTokenPropagation(true,
+                "defaultClientDefaultExchangeOnClassNamedOnMethod_classLevelDefaultClientEnabledExchange");
     }
 
     @Test
     public void testDefaultClientDefaultTokenExchange() {
         testRestClientTokenPropagation(false, "defaultClientDefaultExchange");
+        testRestClientTokenPropagation(false, "defaultClientDefaultExchange_RepeatedAnnotation");
+        testRestClientTokenPropagation(false, "defaultClientDefaultExchangeOnClassNamedOnMethod_classLevelDefaultClient");
         testRestClientTokenPropagation(false, "defaultClientDefaultExchange_OnMethod");
         testRestClientTokenPropagation(false, "multipleClientsAndMultipleMethods_DefaultClientDefaultExchange");
+        testRestClientTokenPropagation(false,
+                "namedClientDefaultExchangeOnClassDefaultClientOnMethod_methodLevelDefaultClientDefaultExchange");
     }
 
     @Test
@@ -93,6 +104,9 @@ public class AccessTokenAnnotationTest {
         testRestClientTokenPropagation(true, "namedClientDefaultExchange");
         testRestClientTokenPropagation(true, "namedClientDefaultExchange_OnMethod");
         testRestClientTokenPropagation(true, "multipleClientsAndMultipleMethods_NamedClientDefaultExchange");
+        testRestClientTokenPropagation(true, "namedClientDefaultExchangeOnClassDefaultClientOnMethod_classLevelNamedClient");
+        testRestClientTokenPropagation(true,
+                "defaultClientDefaultExchangeOnClassNamedOnMethod_classLevelDefaultClientDefaultExchange");
     }
 
     @Test
@@ -117,6 +131,49 @@ public class AccessTokenAnnotationTest {
 
     public String getBearerAccessToken() {
         return client.getAccessToken("alice", "alice");
+    }
+
+    @RegisterRestClient(baseUri = "http://localhost:8081/protected")
+    @AccessToken(exchangeTokenClient = "named")
+    @Path("/")
+    public interface NamedClientDefaultExchangeOnClass_DefaultClientOnMethod {
+
+        @GET
+        String getUserName();
+
+        @AccessToken
+        @GET
+        String getUserName_DefaultClientDefaultExchange();
+
+        @AccessToken(exchangeTokenClient = "Default")
+        @GET
+        String getUserName_DefaultClientEnabledExchange();
+    }
+
+    @RegisterRestClient(baseUri = "http://localhost:8081/protected")
+    @AccessToken
+    @Path("/")
+    public interface DefaultClientDefaultExchangeOnClass_NamedOnMethod {
+
+        @GET
+        String getUserName();
+
+        @AccessToken(exchangeTokenClient = "named")
+        @GET
+        String getUserName_NamedClientDefaultExchange();
+
+        @AccessToken(exchangeTokenClient = "Default")
+        @GET
+        String getUserName_DefaultClientEnabledExchange();
+    }
+
+    @RegisterRestClient(baseUri = "http://localhost:8081/protected")
+    @AccessToken
+    @Path("/")
+    public interface DefaultClientDefaultExchange_RepeatedAnnotation {
+        @AccessToken
+        @GET
+        String getUserName();
     }
 
     @RegisterRestClient(baseUri = "http://localhost:8081/protected")
@@ -187,6 +244,15 @@ public class AccessTokenAnnotationTest {
         String getUserName();
     }
 
+    @AccessToken(exchangeTokenClient = "named")
+    @RegisterRestClient(baseUri = "http://localhost:8081/protected")
+    @Path("/")
+    public interface NamedClientDefaultExchange_RepeatedAnnotation {
+        @AccessToken(exchangeTokenClient = "named")
+        @GET
+        String getUserName();
+    }
+
     // tests no AmbiguousResolutionException is raised
     @Singleton
     @Unremovable
@@ -222,6 +288,22 @@ public class AccessTokenAnnotationTest {
         @Inject
         @RestClient
         MultipleClientsAndMultipleMethods multipleClientsAndMultipleMethods;
+
+        @Inject
+        @RestClient
+        DefaultClientDefaultExchange_RepeatedAnnotation defaultClientDefaultExchange_RepeatedAnnotation;
+
+        @Inject
+        @RestClient
+        NamedClientDefaultExchange_RepeatedAnnotation namedClientDefaultExchange_RepeatedAnnotation;
+
+        @Inject
+        @RestClient
+        NamedClientDefaultExchangeOnClass_DefaultClientOnMethod namedClientDefaultExchangeOnClassDefaultClientOnMethod;
+
+        @Inject
+        @RestClient
+        DefaultClientDefaultExchangeOnClass_NamedOnMethod defaultClientDefaultExchangeOnClassNamedOnMethod;
 
         @Inject
         JsonWebToken jwt;
@@ -280,6 +362,22 @@ public class AccessTokenAnnotationTest {
                     multipleClientsAndMultipleMethods.getUserName_NamedClientDefaultExchange();
                 case "multipleClientsAndMultipleMethods_NoAccessToken" ->
                     multipleClientsAndMultipleMethods.getUserName_NoAccessToken();
+                case "defaultClientDefaultExchange_RepeatedAnnotation" ->
+                    defaultClientDefaultExchange_RepeatedAnnotation.getUserName();
+                case "namedClientDefaultExchange_RepeatedAnnotation" ->
+                    namedClientDefaultExchange_RepeatedAnnotation.getUserName();
+                case "namedClientDefaultExchangeOnClassDefaultClientOnMethod_classLevelNamedClient" ->
+                    namedClientDefaultExchangeOnClassDefaultClientOnMethod.getUserName();
+                case "namedClientDefaultExchangeOnClassDefaultClientOnMethod_methodLevelDefaultClientEnabledExchange" ->
+                    namedClientDefaultExchangeOnClassDefaultClientOnMethod.getUserName_DefaultClientEnabledExchange();
+                case "namedClientDefaultExchangeOnClassDefaultClientOnMethod_methodLevelDefaultClientDefaultExchange" ->
+                    namedClientDefaultExchangeOnClassDefaultClientOnMethod.getUserName_DefaultClientDefaultExchange();
+                case "defaultClientDefaultExchangeOnClassNamedOnMethod_classLevelDefaultClient" ->
+                    defaultClientDefaultExchangeOnClassNamedOnMethod.getUserName();
+                case "defaultClientDefaultExchangeOnClassNamedOnMethod_classLevelDefaultClientEnabledExchange" ->
+                    defaultClientDefaultExchangeOnClassNamedOnMethod.getUserName_DefaultClientEnabledExchange();
+                case "defaultClientDefaultExchangeOnClassNamedOnMethod_classLevelDefaultClientDefaultExchange" ->
+                    defaultClientDefaultExchangeOnClassNamedOnMethod.getUserName_NamedClientDefaultExchange();
                 default -> throw new IllegalArgumentException("Unknown client key");
             };
         }

--- a/extensions/oidc-token-propagation/deployment/src/test/java/io/quarkus/oidc/token/propagation/deployment/test/AccessTokenAnnotationTest.java
+++ b/extensions/oidc-token-propagation/deployment/src/test/java/io/quarkus/oidc/token/propagation/deployment/test/AccessTokenAnnotationTest.java
@@ -38,7 +38,10 @@ public class AccessTokenAnnotationTest {
                             NamedClientDefaultExchange.class, MultiProviderFrontendResource.class, ProtectedResource.class,
                             CustomAccessTokenRequestFilter.class, NamedClientDefaultExchange_OnMethod.class,
                             DefaultClientEnabledExchange_OnMethod.class, DefaultClientDefaultExchange_OnMethod.class,
-                            MultipleClientsAndMultipleMethods.class)
+                            MultipleClientsAndMultipleMethods.class, DefaultClientDefaultExchange_RepeatedAnnotation.class,
+                            NamedClientDefaultExchange_RepeatedAnnotation.class,
+                            DefaultClientDefaultExchangeOnClass_NamedOnMethod.class,
+                            NamedClientDefaultExchangeOnClass_DefaultClientOnMethod.class)
                     .addAsResource(
                             new StringAsset(
                                     """
@@ -76,13 +79,21 @@ public class AccessTokenAnnotationTest {
         testRestClientTokenPropagation(true, "defaultClientEnabledExchange");
         testRestClientTokenPropagation(true, "defaultClientEnabledExchange_OnMethod");
         testRestClientTokenPropagation(true, "multipleClientsAndMultipleMethods_DefaultClientEnabledExchange");
+        testRestClientTokenPropagation(true,
+                "namedClientDefaultExchangeOnClassDefaultClientOnMethod_methodLevelDefaultClientEnabledExchange");
+        testRestClientTokenPropagation(true,
+                "defaultClientDefaultExchangeOnClassNamedOnMethod_classLevelDefaultClientEnabledExchange");
     }
 
     @Test
     public void testDefaultClientDefaultTokenExchange() {
         testRestClientTokenPropagation(false, "defaultClientDefaultExchange");
+        testRestClientTokenPropagation(false, "defaultClientDefaultExchange_RepeatedAnnotation");
+        testRestClientTokenPropagation(false, "defaultClientDefaultExchangeOnClassNamedOnMethod_classLevelDefaultClient");
         testRestClientTokenPropagation(false, "defaultClientDefaultExchange_OnMethod");
         testRestClientTokenPropagation(false, "multipleClientsAndMultipleMethods_DefaultClientDefaultExchange");
+        testRestClientTokenPropagation(false,
+                "namedClientDefaultExchangeOnClassDefaultClientOnMethod_methodLevelDefaultClientDefaultExchange");
     }
 
     @Test
@@ -90,6 +101,9 @@ public class AccessTokenAnnotationTest {
         testRestClientTokenPropagation(true, "namedClientDefaultExchange");
         testRestClientTokenPropagation(true, "namedClientDefaultExchange_OnMethod");
         testRestClientTokenPropagation(true, "multipleClientsAndMultipleMethods_NamedClientDefaultExchange");
+        testRestClientTokenPropagation(true, "namedClientDefaultExchangeOnClassDefaultClientOnMethod_classLevelNamedClient");
+        testRestClientTokenPropagation(true,
+                "defaultClientDefaultExchangeOnClassNamedOnMethod_classLevelDefaultClientDefaultExchange");
     }
 
     @Test
@@ -113,6 +127,49 @@ public class AccessTokenAnnotationTest {
 
     public String getBearerAccessToken() {
         return client.getAccessToken("alice", "alice");
+    }
+
+    @RegisterRestClient(baseUri = "http://localhost:8081/protected")
+    @AccessToken(exchangeTokenClient = "named")
+    @Path("/")
+    public interface NamedClientDefaultExchangeOnClass_DefaultClientOnMethod {
+
+        @GET
+        String getUserName();
+
+        @AccessToken
+        @GET
+        String getUserName_DefaultClientDefaultExchange();
+
+        @AccessToken(exchangeTokenClient = "Default")
+        @GET
+        String getUserName_DefaultClientEnabledExchange();
+    }
+
+    @RegisterRestClient(baseUri = "http://localhost:8081/protected")
+    @AccessToken
+    @Path("/")
+    public interface DefaultClientDefaultExchangeOnClass_NamedOnMethod {
+
+        @GET
+        String getUserName();
+
+        @AccessToken(exchangeTokenClient = "named")
+        @GET
+        String getUserName_NamedClientDefaultExchange();
+
+        @AccessToken(exchangeTokenClient = "Default")
+        @GET
+        String getUserName_DefaultClientEnabledExchange();
+    }
+
+    @RegisterRestClient(baseUri = "http://localhost:8081/protected")
+    @AccessToken
+    @Path("/")
+    public interface DefaultClientDefaultExchange_RepeatedAnnotation {
+        @AccessToken
+        @GET
+        String getUserName();
     }
 
     @RegisterRestClient(baseUri = "http://localhost:8081/protected")
@@ -183,6 +240,15 @@ public class AccessTokenAnnotationTest {
         String getUserName();
     }
 
+    @AccessToken(exchangeTokenClient = "named")
+    @RegisterRestClient(baseUri = "http://localhost:8081/protected")
+    @Path("/")
+    public interface NamedClientDefaultExchange_RepeatedAnnotation {
+        @AccessToken(exchangeTokenClient = "named")
+        @GET
+        String getUserName();
+    }
+
     // tests no AmbiguousResolutionException is raised
     @Singleton
     @Unremovable
@@ -209,6 +275,10 @@ public class AccessTokenAnnotationTest {
 
         @Inject
         @RestClient
+        DefaultClientDefaultExchange_RepeatedAnnotation defaultClientDefaultExchange_RepeatedAnnotation;
+
+        @Inject
+        @RestClient
         DefaultClientEnabledExchange_OnMethod defaultClientEnabledExchange_OnMethod;
 
         @Inject
@@ -217,7 +287,19 @@ public class AccessTokenAnnotationTest {
 
         @Inject
         @RestClient
+        NamedClientDefaultExchange_RepeatedAnnotation namedClientDefaultExchange_RepeatedAnnotation;
+
+        @Inject
+        @RestClient
         MultipleClientsAndMultipleMethods multipleClientsAndMultipleMethods;
+
+        @Inject
+        @RestClient
+        NamedClientDefaultExchangeOnClass_DefaultClientOnMethod namedClientDefaultExchangeOnClassDefaultClientOnMethod;
+
+        @Inject
+        @RestClient
+        DefaultClientDefaultExchangeOnClass_NamedOnMethod defaultClientDefaultExchangeOnClassNamedOnMethod;
 
         @Inject
         JsonWebToken jwt;
@@ -250,8 +332,12 @@ public class AccessTokenAnnotationTest {
                 case "defaultClientEnabledExchange" -> defaultClientEnabledExchange.getUserName();
                 case "namedClientDefaultExchange" -> namedClientDefaultExchange.getUserName();
                 case "defaultClientDefaultExchange_OnMethod" -> defaultClientDefaultExchange_OnMethod.getUserName();
+                case "defaultClientDefaultExchange_RepeatedAnnotation" ->
+                    defaultClientDefaultExchange_RepeatedAnnotation.getUserName();
                 case "defaultClientEnabledExchange_OnMethod" -> defaultClientEnabledExchange_OnMethod.getUserName();
                 case "namedClientDefaultExchange_OnMethod" -> namedClientDefaultExchange_OnMethod.getUserName();
+                case "namedClientDefaultExchange_RepeatedAnnotation" ->
+                    namedClientDefaultExchange_RepeatedAnnotation.getUserName();
                 case "multipleClientsAndMultipleMethods_DefaultClientDefaultExchange" ->
                     multipleClientsAndMultipleMethods.getUserName_DefaultClientDefaultExchange();
                 case "multipleClientsAndMultipleMethods_DefaultClientEnabledExchange" ->
@@ -260,6 +346,18 @@ public class AccessTokenAnnotationTest {
                     multipleClientsAndMultipleMethods.getUserName_NamedClientDefaultExchange();
                 case "multipleClientsAndMultipleMethods_NoAccessToken" ->
                     multipleClientsAndMultipleMethods.getUserName_NoAccessToken();
+                case "namedClientDefaultExchangeOnClassDefaultClientOnMethod_classLevelNamedClient" ->
+                    namedClientDefaultExchangeOnClassDefaultClientOnMethod.getUserName();
+                case "namedClientDefaultExchangeOnClassDefaultClientOnMethod_methodLevelDefaultClientEnabledExchange" ->
+                    namedClientDefaultExchangeOnClassDefaultClientOnMethod.getUserName_DefaultClientEnabledExchange();
+                case "namedClientDefaultExchangeOnClassDefaultClientOnMethod_methodLevelDefaultClientDefaultExchange" ->
+                    namedClientDefaultExchangeOnClassDefaultClientOnMethod.getUserName_DefaultClientDefaultExchange();
+                case "defaultClientDefaultExchangeOnClassNamedOnMethod_classLevelDefaultClient" ->
+                    defaultClientDefaultExchangeOnClassNamedOnMethod.getUserName();
+                case "defaultClientDefaultExchangeOnClassNamedOnMethod_classLevelDefaultClientEnabledExchange" ->
+                    defaultClientDefaultExchangeOnClassNamedOnMethod.getUserName_DefaultClientEnabledExchange();
+                case "defaultClientDefaultExchangeOnClassNamedOnMethod_classLevelDefaultClientDefaultExchange" ->
+                    defaultClientDefaultExchangeOnClassNamedOnMethod.getUserName_NamedClientDefaultExchange();
                 default -> throw new IllegalArgumentException("Unknown client key");
             };
         }

